packages/pubsub/src/subscription.js

```diff
@@ -403,84 +234,51 @@ Subscription.formatName_ = function(projectId, name) {
 };
 
 /**
- * Generate a random name to use for a name-less subscription.
- *
- * @private
- */
-Subscription.generateName_ = function() {
-  return 'autogenerated-' + uuid.v4();
-};
-
-/**
- * Acknowledge to the backend that the message was retrieved. You must provide
- * either a single ackId or an array of ackIds.
- *
- * @resource [Subscriptions: acknowledge API Documentation]{@link https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/acknowledge}
- *
- * @throws {Error} If at least one ackId is not provided.
- *
- * @param {string|string[]} ackIds - An ackId or array of ackIds.
- * @param {options=} options - Configuration object.
- * @param {number} options.timeout - Set a maximum amount of time in
- *     milliseconds before giving up if no response is received.
- * @param {function=} callback - The callback function.
- *
- * @example
- * var ackId = 'ePHEESyhuE8e...';
- *
- * subscription.ack(ackId, function(err, apiResponse) {});
- *
- * //-
- * // If the callback is omitted, we'll return a Promise.
- * //-
- * subscription.ack(ackId).then(function(data) {
- *   var apiResponse = data[0];
- * });
+ * This should never be called directly.
  */
-Subscription.prototype.ack = function(ackIds, options, callback) {
+Subscription.prototype.ack_ = function(messages, callback) {
   var self = this;
-
-  ackIds = arrify(ackIds);
-
-  if (ackIds.length === 0) {
-    throw new Error([
-      'At least one ID must be specified before it can be acknowledged.'
-    ].join(''));
-  }
-
-  if (is.fn(options)) {
-    callback = options;
-    options = {};
-  }
-
-  options = options || {};
-  callback = callback || common.util.noop;
-
-  var protoOpts = {
-    service: 'Subscriber',
-    method: 'acknowledge'
-  };
-
-  if (options && is.number(options.timeout)) {
-    protoOpts.timeout = options.timeout;
+  var ackIds = messages.reduce(function(ackIds, message) {
+    ackIds.push(message.ackId);
+    return ackIds;
+  }, []);
+
+  if (this.connection) {
+    this.connection.write({ ackIds: ackIds });
+    setImmediate(onComplete);
+    return;
   }
 
   var reqOpts = {
-    subscription: this.name,
-    ackIds: ackIds
+    ackIds: ackIds,
+    subscription: this.name
   };
 
-  this.parent.request(protoOpts, reqOpts, function(err, resp) {
-    if (!err) {
-      ackIds.forEach(function(ackId) {
-        delete self.inProgressAckIds[ackId];
-      });
+  this.request({
+    client: 'subscriberClient',
+    method: 'acknowledge',
+    reqOpts: reqOpts
+  }, onComplete);
+
+  function onComplete(err, resp) {
+    messages.forEach(function(message) {
+      self.messages_.delete(message);
+    });
 
-      self.refreshPausedStatus_();
+    if (self.connection.isPaused()) {
+      self.connection.resume();
```

Comment:
Gotcha, I went ahead and made [some changes](https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2380/files#diff-1f00668470ede3141cbf2a3dcaf4a60bR257) so now we'll resume whenever an ack/nack occurs.